pkg/docker/util.go

```diff
@@ -272,64 +272,87 @@ func PullImage(name string, d Docker, policy api.PullPolicy) (*PullResult, error
 	return &PullResult{Image: image, OnBuild: d.IsImageOnBuild(name)}, err
 }
 
-// CheckAllowedUser retrieves the user for a Docker image and checks that user
-// against an allowed range of uids.
+// CheckAllowedUser retrieves the execution users for a Docker image and
+// checks that user against an allowed range of uids.
 // - If the range of users is not empty, then the user on the Docker image
 // needs to be a numeric user
 // - The user's uid must be contained by the range(s) specified by the uids
 // Rangelist
+// - If build image uses an assemble user (via a command override or an
+// image label), that user must be within the allowed range of uids.
 // - If the image contains ONBUILD instructions and those instructions also
-// contain a USER directive, then the user specified by that USER directive
+// contain any USER directives, then all users specified by those USER directives
 // must meet the uid range criteria as well.
-func CheckAllowedUser(d Docker, imageName string, uids user.RangeList, isOnbuild bool) error {
+func CheckAllowedUser(d Docker, imageName string, uids user.RangeList, isOnbuild bool, assembleUserOverride string) error {
 	if uids == nil || uids.Empty() {
 		return nil
 	}
-	imageUserSpec, err := d.GetImageUser(imageName)
+	assembleUser, err := extractAssembleUser(d, assembleUserOverride, imageName)
+	if err != nil {
+		return err
+	}
+	if len(assembleUser) > 0 {
+		if !user.IsUserAllowed(assembleUser, &uids) {
```

Comment:
if the assembleUser is allowed, I don't think we need to proceed to check if the imageUser is allowed?

If anything we probably don't want to because if someone has an image that runs as root, but they've explicitly said "i want to run my assemble script as $randomuser" then we should allow that because it's safe.

the on-build case is trickier since in that case the USER statement will get executed regardless of any overrides.